fem/libceed/pa_integrator.hpp

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_LIBCEED_PAINTEG
+#define MFEM_LIBCEED_PAINTEG
+
+#include "mf_integrator.hpp"
+#include "../../config/config.hpp"
+#include "../fespace.hpp"
+#include "../gridfunc.hpp"
+
+namespace mfem
+{
+
+namespace ceed
+{
+
+#ifdef MFEM_USE_CEED
+/** This structure contains the data to assemble a PA operator with libCEED.
+    See libceed/mass.cpp or libceed/diffusion.cpp for examples. */
+struct PAOperator
+{
+   /** The finite element space for the trial and test functions. */
+   const mfem::FiniteElementSpace &fes;
+   /** The Integration Rule to use to compute the operator. */
+   const mfem::IntegrationRule &ir;
+   /** The number of quadrature data at each quadrature point. */
+   int qdatasize;
+   /** The path to the header containing the functions for libCEED. */
+   std::string header;
+   /** The name of the Qfunction to build the quadrature data. */
+   std::string build_func;
+   /** The Qfunction to build the quadrature data. */
+   CeedQFunctionUser build_qf;
+   /** The name of the Qfunction to apply the operator. */
+   std::string apply_func;
+   /** The Qfunction to apply the operator. */
+   CeedQFunctionUser apply_qf;
+   /** The evaluation mode to apply to the trial function (CEED_EVAL_INTERP,
+       CEED_EVAL_GRAD, etc.) */
+   EvalMode trial_op;
+   /** The evaluation mode to apply to the test function ( CEED_EVAL_INTERP,
+       CEED_EVAL_GRAD, etc.)*/
+   EvalMode test_op;
+};
+#endif
+
+class PAIntegrator : public MFIntegrator
+{
+#ifdef MFEM_USE_CEED
+protected:
+   CeedOperator build_oper;
+   CeedQFunction build_qfunc;
+
+public:
+   PAIntegrator()
+      : MFIntegrator(),  build_oper(nullptr), build_qfunc(nullptr) { }
+
+   template <typename CeedInfo, typename CoeffType>
+   PAOperator InitPA(CeedInfo &info,
```

Comment:
This could use a few lines of documentation; also using the template parameter `CeedKernelInfo` IMO would make it more readable.

Reply:
I think I could (should?) get rid of this method and the `PAOperator` struct (and `MFOperator` struct) and change the `Assemble` method's signature.

Reply:
I went ahead and removed `InitPA` and `InitMF` and fused them with `Assemble`, it's indeed more clean and clear. Thank you for the suggestion.

Reply:
In order to make the documentation more clear I introduced an `OperatorInfo` struct from which the template parameter `CeedInfo` is supposed to inherit (see `mass.cpp` for instance), I think the design is even more clean and clear now.